### DISCUSSION TITLE
Storage usage

### DISCUSSION THREAD

**Discussion by tobiasdiez** [community/community#38920] [Codespaces]
### Select Topic Area

Product Feedback

### Body

A few suggestions for improvement concerning the storage of codespaces:
- In the overview https://github.com/codespaces, each codespace should show how much storage it consumes.
- If you delete a codespace, it doesn't seem to decrease the amount if storage consumed, or at least it takes quite some time until it is updated at https://github.com/settings/billing. This is inconvenient if you run out of your (free) storage and delete some old codespace to continue working, but that doesn't work. Edit: apparently, github uses the unit "GB - month" so it's not the total storage used but also takes into account how long you use that storage.
- Under https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#billing-for-storage-usage, you find "When you use the default dev container configuration (see "[Introduction to dev containers](https://docs.github.com/en/codespaces/setting-up-your-project-for-codespaces/introduction-to-dev-containers#using-the-default-dev-container-configuration)"), we do not count the default container as used storage.". I think this sets the wrong incentive as people are not rewarded to optimize the consumed storage by using one of the other smaller base images, that only include the features actually used. I would suggest to exempt all of the official images at https://hub.docker.com/_/microsoft-vscode-devcontainers from counting towards the used storage.

**Comment by EndBug** [comment]
I would also add a way to better explore storage usage, in particular regarding **prebuilds**: I almost immediately ran out of this month's free storage usage because I had set up some prebuilds and forgot about them.
As of now, there's no way to see which repos have prebuilds on github.com, the only way is downloading the usage report and seeing what repos are you being billed for.

**Reply by AngelOnFira** [reply]
Bumping this. As I'm setting up more prebuilds on repos I use for teaching others, I'm occurring higher costs quite quickly. I'm not making nearly as much use of my core-hours as I am of storage it seems. More breakdowns would help a lot.

**Comment by cubxxw** [comment]
我很无奈，我用的很快，没人提醒我，我很赞同你说的，The margin should be displayed

**Comment by DanielHabenicht** [comment]
I can only add my support, on the first day of non-beta I ran out of space too. And it was not apparent where it came from.

**Comment by Josverl** [comment]
same issue here, ran out in a matter of days over the weekend., apparently storage is billed per minute even if codespaces is not used ?
which for me would make it impractical/impossible to use it for my OSS contributions which by nature require a discontinuous use rhythm alternating work and OSS

**Reply by DanielHabenicht** [reply]
Yes, I was wondering about that too. I am contributing to jitsi-meet and there Codespace uses `0.4512` (daily!) and my pull request is open for more than 2 months now... which means I can work on 2 repositories concurrently if I don't want to halt my works in the middle of the month. 

@ github Maybe it is possible to only charge for the last layer (of user changes, e.g. Environment variables, extra files etc.) and delete the rest of the docker container. I am more than happy to wait for a codespace for half a minute while it gets build again - while not being charged for that much storage.

**Reply by samkio** [reply]
The codespace doesn't have to be there for the whole time a PR is open however right? You can shutdown/delete the codespace once the change is committed and re-spin up a new codespace if changes are required etc.

It does negate one of the best features of codespaces IMHO however; being able to continue where you left off with the state exactly as you left it was great. But with this restriction it appears we can no longer do this without limiting the number of active codespaces. :(

**Reply by DanielHabenicht** [reply]
Yeah but its more work, and currently this workflow is not really supported by the tools so you always have to delete it manually from the Codespaces Tab. 

E.g. There is no option to delete the workspace after you are done (from UI/VSCode/etc.), no automatic deletion.

The second part you mentioned is also the most important part for me. 
Most software projects need some kind of configuration, wich is not tracked in git and the chances are very high that for developing a feature you need to set them up. 
So one pretty nice feature would be to just save those changes in the background and apply them after a rebuild (so you can delete the workspace while you dont need it).

**Comment by craiglpeters** [comment]
What you are saying about incentives makes sense... we have a few different levers which we can pull, and are weighing the various tradeoffs.

[Codespaces has two storage components](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#billing-for-storage-usage):

1. Codespaces provides persistence for work whether the codespace is active or suspended. We count storage on an hourly basis while a codespace exists, so the used storage is cumulative during a billing period. This means that deleting storage (or a codespace) reduces the rate of storage consumption. But if you have one or more codespaces you will be accumulating storage consumption.
2. Storage for [Codespaces prebuilds](https://docs.github.com/en/codespaces/prebuilding-your-codespaces/about-github-codespaces-prebuilds) enables faster codespace starts for large repositories or complex container builds

We are continually working to improve the experience and the following are in progress

- We will display user storage for each codespace on github.com/codespaces soon to help you understand which codespaces are using how much storage - stay tuned
- If your project is open source, you can apply for a limited prebuild storage grants. See https://github.co/prebuilds_for_oss

Note that you may have codespaces that have been around since before the [retention period for Codespaces](https://docs.github.com/en/codespaces/managing-codespaces-for-your-organization/restricting-the-retention-period-for-codespaces) was implemented. You can [export work from a codespace to a branch](https://docs.github.com/en/codespaces/troubleshooting/exporting-changes-to-a-branch) at any time so there is no need to keep a codespace around longer than the task at hand.

**Comment by treeder** [comment]
I would like to know which codespaces are using storage too. All of a sudden (I presume now that you are charging for personal accounts) I hit my 20GB Pro Plan limit on Codespaces storage, but I have no idea how much each of my Codespaces is using. So my only option is to increase my budget since I don't know which Codespaces to terminate.

**Reply by DanielHabenicht** [reply]
For now you can go to https://github.com/settings/billing and request the billing report: 
![image](https://user-images.githubusercontent.com/13590797/202462470-c8ddba9a-4c5c-400e-b6da-a6d6857665c8.png)

**Reply by mikeando** [reply]
As an example row for a pretty default rust codespace I have:

| Date | Product | SKU | Quantity | Unit Type | Price Per Unit ($)| Multiplier |
| ---- | ------ | --- | ------ | --------- | -----| ----- |
| 2022-11-13 | Codespaces - Linux | Storage | 1.032 | gb-month | 0.07| 1.0 |

Is this telling me that I'm using 1gb-month every day in that code space?
That would presumably mean that the code-space size is 30gb?
That seems excessively large, and tricky to diagnose.

**Comment by RyanLangston** [comment]
> * Under https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#billing-for-storage-usage, you find "When you use the default dev container configuration (see "[Introduction to dev containers](https://docs.github.com/en/codespaces/setting-up-your-project-for-codespaces/introduction-to-dev-containers#using-the-default-dev-container-configuration)"), we do not count the default container as used storage.". I think this sets the wrong incentive as people are not rewarded to optimize the consumed storage by using one of the other smaller base images, that only include the features actually used. I would suggest to exempt all of the official images at https://hub.docker.com/_/microsoft-vscode-devcontainers from counting towards the used storage.

This one seems especially important
